app/demo-loop/LDemoKernel.cu

```diff
@@ -0,0 +1,67 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file LDemoKernel.cu
+//---------------------------------------------------------------------------//
+#include "LDemoKernel.hh"
+
+#include "base/KernelParamCalculator.cuda.hh"
+#include "random/cuda/RngEngine.hh"
+#include "random/distributions/ExponentialDistribution.hh"
+#include "physics/base/PhysicsStepUtils.hh"
+
+using namespace celeritas;
+
+namespace demo_loop
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+// KERNELS
+//---------------------------------------------------------------------------//
+__global__ void
+pre_step_kernel(ParamsDeviceRef const params, StateDeviceRef const states)
+{
+    auto tid = celeritas::KernelParamCalculator::thread_id();
+    if (tid.get() >= states.size())
+        return;
+
+    CELER_NOT_IMPLEMENTED("Pre-step kernel");
+#if 0
+    PhysicsTrackView phys(inp.params, inp.states, init.particle, init.mat, tid);
+    phys = PhysicsTrackInitializer{};
+
+    // Sample mean free path
+    {
+        RngEngine                 rng(states.rng, ThreadId(tid));
+        ExponentialDistribution<> sample_exponential;
+        phys.interaction_mfp(sample_exponential(rng));
+    }
+#endif
+}
+
+} // namespace
+//---------------------------------------------------------------------------//
+// KERNEL INTERFACES
+//---------------------------------------------------------------------------//
+#define CDL_LAUNCH_KERNEL(NAME, THREADS, ARGS...)                            \
+    do                                                                       \
+    {                                                                        \
+        static const ::celeritas::KernelParamCalculator calc_kernel_params_( \
+            NAME##_kernel, #NAME);                                           \
+        auto grid_ = calc_kernel_params_(THREADS);                           \
```

Comment:
```suggestion
        auto grid = calc_kernel_params_(THREADS);                           \
```
isn't underscore for data member?

Reply:
Yeah, I put the suffix here to indicate that it's sort of a "private scope" to avoid conflicting with anything else in the parent -- I guess I could concatenate as `NAME##_grid` to do that? There's similar code in the local variables inside `base/Assert.hh` as well.

Reply:
> I guess I could concatenate as NAME##_grid

This would be better albeit possibly 'overkill' since there is already a scope, all we need is a symbol that is that can not 'hide' any of the variable used in the parameters.  Humm ... `cdl_launch_kernel_grid` is an alternative ... then `NAME##_grid` probably actually the best choice.

Also why the implied required:
``` NAME##_kernel<<<```
of the function name having the _kernel suffix?   Wouldn't be simpler to just use `(NAME, #NAME)`?
Which would also have the advantage to record the full function name in the diagnostic ...

Reply:
I've changed the `calc_kernel_params` to include the name of the kernel, since it's static data and thus the name will show up in the binary (if symbols are exported).

As to the `_kernel` extension, this is just following the convention of "function X launches kernel function X_kernel and saves the debug data to X". Since all the functions saved to the kernel diagnostics (which is where the string `#NAME` is used) are kernel functions, I'd rather them all just show up as keys without "_kernel" ;)

Reply:
Fair enough :)